File: SetOfStacks.java
import java.util.ArrayList;
import java.util.EmptyStackException;
class Node{
  public Node above, below;
  public int value;
  public Node(int value){
    this.value = value;
  }
}
public class SetOfStacks{
  private ArrayList<Stack> stacks = new ArrayList<Stack>();
  private int capacity;
  public SetOfStacks(int capacity){
    this.capacity = capacity;
  }

  public Stack getLastStack(){
    if (stacks.size() == 0) return null;
    return stacks.get(stacks.size()-1);
  }

  public void push(int value){
    Stack last = getLastStack();
    if (last != null && !last.isFull()){
      last.push(value);
    } else {
      Stack stack =  new Stack(capacity);
      stack.push(value);
      stacks.add(stack);
    }
  }
  public int pop(){
    Stack last = getLastStack();
    if (last == null){
      throw new EmptyStackException();
    }
    int value = last.pop();
    if (last.size == 0) stacks.remove(stacks.size()-1);
    return value;
  }

  public boolean isEmpty(){
    Stack last = getLastStack();
    return last == null || last.isEmpty();
  }

  public int popAt(int index){
    if (index < 0 || index > stacks.size()) throw new IndexOutOfBoundsException();
    return leftShift(index, true);
  }

  public int leftShift(int index, boolean removeTop){
    Stack stack = stacks.get(index);
    int removed_item;
    if (removeTop) removed_item = stack.pop();
    else removed_item = stack.removeBottom();
    if (stacks.get(index).isEmpty()){
      stacks.remove(index);
    } else if (index + 1 < stacks.size()){
      int value = leftShift(index+1, false);
      stack.push(value);
    }
    return removed_item;
  }
}
class Stack{
  private int capacity;
  public Node top, bottom;
  public int size = 0;

  public Stack(int capacity){
    this.capacity = capacity;
  }
  public boolean isFull(){ return capacity == size;}
  public boolean isEmpty(){ return size == 0;}

  public void join(Node above, Node below){
    if (below != null) below.above = above;
    if (above != null) above.below = below;
  }
  public boolean push(int value){
    Node n = new Node(value);
    if (size < capacity){
      join(n,top);
      top = n;
      if (size == 1) bottom = n;
      size ++;
      return true;
    } else {
      return false;
    }
  }

  public int pop(){
    if (size == 0) throw new EmptyStackException();
    Node t = top;
    top = top.below;
    if (top != null) top.above = null;
    if (top == null) bottom = null;
    size --;
    return t.value;
  }

  public int removeBottom(){
    Node b = bottom;
    bottom = bottom.above;
    if (bottom != null) bottom.below = null;
    size --;
    return b.value;
  }
}

File: SetOfStacks.py
class Node():
    def __init__(self,value):
        self.value = value
        self.above = None
        self.below = None

class Stack():
    def __init__(self,capacity):
        self.capacity = capacity
        self.top = None
        self.bottom = None
        self.size = 0

    def isEmpty(self):
        print(self.size)
        return self.size == 0
    def isFull(self):
        return self.size == self.capacity

    def join(self, above, below):
        if above != None:
            above.below = below
        if below != None:
            below.above = above

    def push(self, value):
        n = Node(value)
        if self.size < self.capacity:
            self.join(n, self.top)
            self.top = n
            if self.size == 0:
                self.bottom = n
            self.size += 1
            return True
        else:
            return False

    def pop(self):
        if self.size == 0:
            raise Exception("This Stack is empty fool")
        else:
            value = self.top.value
            self.top = self.top.below
            if self.top == None:
                self.bottom = None
            else:
                self.top.above = None
            self.size -= 1
            return value

    def removeBottom(self):
        b = self.bottom
        self.bottom = b.above
        if self.bottom != None:
            self.bottom.below = None
        self.size -= 1
        return b.value

    def printStack(self):
        curr = self.bottom
        lst = []
        while curr != None:
            lst.append(curr.value)
            curr = curr.above
        print(lst)

class SetOfStacks():
    def __init__(self, capacity):
        self.__list = []
        self.__capacity = capacity

    def getLastStack(self):
        if len(self.__list) == 0:
            return None
        return self.__list[-1]

    def push(self, value):
        last = self.getLastStack()
        if last == None or last.isFull():
            last = Stack(self.__capacity)
            self.__list.append(last)
        last.push(value)

    def pop(self):
        last = self.getLastStack()
        if last == None or last.isEmpty():
            raise Exception("This set has no stack")
        else:
            value = last.pop()
            if last.isEmpty():
                self.__list.remove(last)
            return value

    def popAt(self, index):
        if index<0 or index>=len(self.__list):
            raise Exception("Index out of bounds")
        return self.leftShift(index, True)

    def leftShift(self, index, removeTop):
        removed_item = None
        stack = self.__list[index]
        if removeTop:
            removed_item = stack.pop()
        else:
            removed_item = stack.removeBottom()
        if stack.isEmpty():
            print("Not this")
            self.__list.remove(stack)
        elif index+1 < len(self.__list):
            newTop = self.leftShift(index+1,False)
            stack.push(newTop)
        return removed_item

    def isEmpty(self):
        if len(self.__list) == 0:
            return True
        return False

    def printSet(self):
        print(self.__list)
        for i in range(len(self.__list)):
            print(i)
            print(self.__list[i].printStack())
